src/main/scala/diplomacy/Nodes.scala

```diff
@@ -5,123 +5,328 @@ package freechips.rocketchip.diplomacy
+  * @tparam B  Bundle type generated by connections to this node
+  */
 abstract class NodeImp[D, U, EO, EI, B <: Data]
   extends Object with InwardNodeImp[D, U, EI, B] with OutwardNodeImp[D, U, EO, B]
 
-// If your edges have the same direction, using this saves you some typing
+/** A [[NodeImp]] where the inward and outward edges are of the same type.
+  *
+  * If the parameters on inward flowing edges are the same as parameters on outer flowing edges, the [[SimpleNodeImp]] can be used.
+  *
+  * @tparam D type of the Downwards flowing parameters of the node
+  * @tparam U type of the Upwards flowing parameters of the node
+  * @tparam E Edge Parameters describing the connections on either side of the node
+  * @tparam B Bundle type generated on edges connecting to this node.
+  *
+  *           If your edges have the same type in both directions, using this saves you some typing.
+  */
 abstract class SimpleNodeImp[D, U, E, B <: Data]
   extends NodeImp[D, U, E, E, B]
 {
+  /** Create an edge that that passes through this node.
+    * 
+    * @param pd downwards flowing parameters along this edge
+    * @param pu upwards flowing parameters of this edge
+    * @param p [[Parameter]]s which can be used during negotiation
+    * @param sourceInfo [[SourceInfo]] of this edge.
+    * @return negotiated edge parameters
+    */
   def edge(pd: D, pu: U, p: Parameters, sourceInfo: SourceInfo): E
-  def edgeO(pd: D, pu: U, p: Parameters, sourceInfo: SourceInfo) = edge(pd, pu, p, sourceInfo)
-  def edgeI(pd: D, pu: U, p: Parameters, sourceInfo: SourceInfo) = edge(pd, pu, p, sourceInfo)
+  /** Create an edge coming out of this node from this node.
+    * 
+    * By default just calls [[edge]].
+    */
+  def edgeO(pd: D, pu: U, p: Parameters, sourceInfo: SourceInfo): E = edge(pd, pu, p, sourceInfo)
+  /** Create an input edge into this node.
+    * 
+    *  By default just calls [[edge]].
+    */
+  def edgeI(pd: D, pu: U, p: Parameters, sourceInfo: SourceInfo): E = edge(pd, pu, p, sourceInfo)
+
+  /** Generate the Bundle from the negotiated Edge parameters.
+    * 
+    * @param e the negotiated Edge parameters
+    * @return the corresponding Bundle of this node
+    */
   def bundle(e: E): B
-  def bundleO(e: E) = bundle(e)
-  def bundleI(e: E) = bundle(e)
+  /** Generate an output Bundle from the negotiated Edge parameters.
+    * 
+    *  By default just calls [[bundle]].
+    */
+  def bundleO(e: E): B = bundle(e)
+  /** Generate an input Bundle from the negotiated Edge parameters.
+    * 
+    *  By default just calls [[bundle]].
+    */
+  def bundleI(e: E): B = bundle(e)
 }
 
+/** [[BaseNode]] is the abstract base class of diplomacy node system.
+  * 
+  *  While [[*NodeImp]] associates the different types in a diplomatic
+  *  system, the [[BaseNode]] uses the parameter types to build
+  *  actual nodes and connect them with edges in the graph through binding
+  *  operations.
```

Comment:
I would not say that BaseNode uses anything. Rather, sometimes one wants to view the entire diplomacy graph in a way where you do not care about the specific types of the edges. BaseNodes are type-erased and provide this view.